### PR TITLE
Query Trace Recording

### PR DESCRIPTION
**Pull request by rabbit721**
This PR adds new infrastructure to record SQL queries #1113.

### CODE REVIEW

File: src/traffic_cop/traffic_cop.cpp [hunk 1]
```diff
@@ -364,6 +366,16 @@ TrafficCopResult TrafficCop::CodegenPhysicalPlan(
...
 
   // TODO(Matt): handle code generation failing
+  auto &resource_metrics = common::thread_context.resource_tracker_.GetMetrics();
+
+  const bool query_trace_metrics_enabled =
+      common::thread_context.metrics_store_ != nullptr &&
+      common::thread_context.metrics_store_->ComponentToRecord(metrics::MetricsComponent::QUERY_TRACE);
+  if (query_trace_metrics_enabled) {
+    common::thread_context.metrics_store_->RecordQueryText(exec_query->GetQueryId(),
+                                                           portal->GetStatement()->GetQueryText(), resource_metrics);
```
Comment (malin1993ml) [inline src/traffic_cop/traffic_cop.cpp:376]: We don't need the `resource_metrics` for query texts. We can just remove that one. Instead, we can append the timestamp at the end.

File: src/traffic_cop/traffic_cop.cpp [hunk 2]
```diff
@@ -406,6 +418,17 @@ TrafficCopResult TrafficCop::RunExecutableQuery(const common::ManagedPointer<net
...
 
+  auto &resource_metrics = common::thread_context.resource_tracker_.GetMetrics();
+
+  const bool query_trace_metrics_enabled =
+      common::thread_context.metrics_store_ != nullptr &&
+      common::thread_context.metrics_store_->ComponentToRecord(metrics::MetricsComponent::QUERY_TRACE);
+
+  if (query_trace_metrics_enabled) {
+    common::thread_context.metrics_store_->RecordQueryTrace(exec_query->GetQueryId(), metrics::MetricsUtil::Now(),
+                                                            resource_metrics);
```
Comment (malin1993ml) [inline src/traffic_cop/traffic_cop.cpp:429]: Similarly, we don't need the `resource_metrics` here as well.

**Review by malin1993ml — CHANGES_REQUESTED** [review]
I think overall this is good! Just two minor changes.

A high-level thing is that I think we also need to add a test case for the actual data recording like [here](https://github.com/cmu-db/terrier/blob/master/test/metrics/metrics_test.cpp#L74). For that, you can reference the [traffic_cop_test](https://github.com/cmu-db/terrier/blob/master/test/traffic_cop/traffic_cop_test.cpp). You pretty much need to [enable all the layers](https://github.com/cmu-db/terrier/blob/master/test/traffic_cop/traffic_cop_test.cpp#L41) of the system. Then create a table and issue some simple queries to the system. The query id and the timestamp might be difficult to check, but at least you can check that the query string is correct and the right number of records is recorded. The `create table` query may not be logged but that's okay.

File: src/traffic_cop/traffic_cop.cpp (hunk 1 above)
Comment (rabbit721) [reply src/traffic_cop/traffic_cop.cpp]: Removed `resource_metrics`.

File: src/traffic_cop/traffic_cop.cpp (hunk 2 above)
Comment (rabbit721) [reply src/traffic_cop/traffic_cop.cpp]: Removed `resource_metrics`.

File: src/traffic_cop/traffic_cop.cpp [hunk 3]
```diff
@@ -364,6 +366,15 @@ TrafficCopResult TrafficCop::CodegenPhysicalPlan(
       common::ManagedPointer<const std::string>(&portal->GetStatement()->GetQueryText()));
 
   // TODO(Matt): handle code generation failing
+
+  const bool query_trace_metrics_enabled =
+      common::thread_context.metrics_store_ != nullptr &&
+      common::thread_context.metrics_store_->ComponentToRecord(metrics::MetricsComponent::QUERY_TRACE);
+  if (query_trace_metrics_enabled) {
+    common::thread_context.metrics_store_->RecordQueryText(exec_query->GetQueryId(),
+                                                           portal->GetStatement()->GetQueryText());
```
Comment (malin1993ml) [inline src/traffic_cop/traffic_cop.cpp:375]: Let's add the timestamp as the third parameter here. May also be useful.

File: test/metrics/metrics_test.cpp [hunk 4]
```diff
@@ -199,6 +213,56 @@ TEST_F(MetricsTests, TransactionCSVTest) {
...
+    EXPECT_TRUE(false);
+  }
+
+  std::this_thread::sleep_for(std::chrono::seconds(1));
+
+  metrics_manager_->Aggregate();
+  const auto aggregated_data = reinterpret_cast<QueryTraceMetricRawData *>(
+      metrics_manager_->AggregatedMetrics().at(static_cast<uint8_t>(MetricsComponent::QUERY_TRACE)).get());
+  EXPECT_NE(aggregated_data, nullptr);
+  EXPECT_GE(aggregated_data->query_trace_.size(), 2);  // 2 data point recorded
```
Comment (malin1993ml) [inline test/metrics/metrics_test.cpp:248]: Should this be `EXPECT_GE` or `EXPECT_EQ`?

File: test/metrics/metrics_test.cpp [hunk 5]
```diff
@@ -199,6 +213,56 @@ TEST_F(MetricsTests, TransactionCSVTest) {
...
+  if (!(aggregated_data->query_trace_.empty())) {
+    EXPECT_GE(aggregated_data->query_trace_.begin()->timestamp_, 2);  // 1 records: create table
```
Comment (malin1993ml) [inline test/metrics/metrics_test.cpp:250]: Same here. Also, is the "create table" comment valid here?

File: test/metrics/metrics_test.cpp [hunk 6]
```diff
@@ -199,6 +213,56 @@ TEST_F(MetricsTests, TransactionCSVTest) {
...
+  }
+  EXPECT_GE(aggregated_data->query_text_.size(), 2);  // 2 data point recorded
```
Comment (malin1993ml) [inline test/metrics/metrics_test.cpp:252]: Same here.

**Review by malin1993ml — CHANGES_REQUESTED** [review]
Thanks for the updates! Only a few nits left. Also, you need to merge the latest master and resolve the conflicts.

File: test/metrics/metrics_test.cpp (hunk 4 above)
Comment (rabbit721) [reply test/metrics/metrics_test.cpp]: Fixed it to be `EXPECT_EQ`.

**Review by malin1993ml — APPROVED** [review]
LGTM! Just need to pass the build.

**Pull request merged**